### ISSUE TITLE
[PS] Reconsider the use of "dummy" check boxes in the permissions page.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5907]
This issue here was triggered after research and reviewing the various `max_input_vars`-related issues in the Backdrop (see #3824 / #5771 etc.) and Drupal core issue queues.

From `core/modules/user/js/user.permissions.js`:

```php
      // Create dummy checkboxes. We use dummy checkboxes instead of reusing
      // the existing checkboxes here because new checkboxes don't alter the
      // submitted form. If we'd automatically check existing checkboxes, the
      // permission table would be polluted with redundant entries. This
      // is deliberate, but desirable when we automatically check them.
```

This is something we've inherited from drupal. I would like to get a better understanding of:
- why we are not using `#states` instead of dummy checkboxes
- what "polluted with redundant entries" means, and why can't that be sorted in the form submit handler, or in the validation handler (which is called before the submit handler)

The reason why I have opened this issue here is because I don't think that when these dummy checkboxes were initially implemented in Drupal the issue with `max_input_vars` and memory limit had surfaced yet, so that serious side effect was not considered. Consider this: on a site with enough roles and permissions to reach 2000 checkboxes on that page, we are adding another 2000 "dummy" checkboxes on top of that, which makes the total number of inputs on the page double ...that's a serious overhead, and even with the limit set to 10000, we still get only 5000 inputs effectively available - the other 5000 half is just useless placeholders (well, roughly speaking - there are no dummy checkboxes for the checkboxes for the anonymous and authenticated roles, so these numbers actually increase with the number of other roles). I know that these dummy checkboxes have been introduced for a reason (they do have a purpose) - what I am wondering is if we can come up with an alternative solution now that we know more/better.

This that I have done some initial research for:
- placing `<input>` checkboxes outside the `<form>` element, "compiling" a list of less elements (only the checked checkboxes), and finally submitting only those.
- creating separate `<form>`s (say one for each module and the permissions provided by it), and then triggering multiple form submissions via a single submit button and JavaScript/AJAX (see `Backdrop.ajax.prototype.beforeSubmit` in `core/misc/ajax.js`).
- adding a `'changed'` class to checkboxes that have been ticked/unticked (via JavaScript), and then only submitting those, discarding any unchanged ones.
- adding a hidden form field of `'#type' => 'value'`, and storing there the existing permissions, then comparing the form to that in order to determine what has changed.
- use cookies or localStorage/sessionStorage instead of the dummy checkboxes
- use `data-*` attributes instead of the dummy checkboxes

**Comment by klonos** [comment]
...or could it be as simple as running a JS function on submit that simply removes all the dummy checkboxes before the actual submit of the form (so that they don't even get posted in the first place)? 🤔

**Comment by kiamlaluno** [comment]
That code has been introduced in Drupal as per [Usability: Auto-check permissions if "authenticated" has them](https://www.drupal.org/project/drupal/issues/78941), an issue created and fixed for Drupal 6, which doesn't have states. That code has never been changed, after Drupal 6.

As for why states aren't used, [the code before that part](https://github.com/backdrop/backdrop/blob/1.x/core/modules/user/js/user.permissions.js#L10) should explain why they aren't used.

```js
// On a site with many roles and permissions, this behavior initially has
// to perform thousands of DOM manipulations to inject checkboxes and hide
// them. By detaching the table from the DOM, all operations can be
// performed without triggering internal layout and re-rendering processes
// in the browser.
var $table = $(this);
var $ancestor, method;
if ($table.prev().length) {
  $ancestor = $table.prev();
  method = 'after';
}
else {
  $ancestor = $table.parent();
  method = 'append';
}
$table.detach();
```

From that comment, I take states aren't used because the JavaScript code that disables the checkboxes for those permissions that are inherited from the authenticated role needs to first detach the permission table and then disable all the checkboxes. Using states, each permission row would have its own state that would cause re-rendering of every permission row.

I don't know if with the browsers supported by Backdrop that would be a problem. Do browser try to render the part of the page that isn't visible? What happens, if states would be used instead of that JavaScript code?

**Comment by klonos** [comment]
Thanks for the additional research and background @kiamlaluno 🙏🏼 ...I don't have the answers to all questions - all I'm saying is that for every permission checkbox other than anonymous and authenticated, we are adding 2 checkboxes in the form, when we only really need one. That is silly given the issue with `max_input_vars`, and it makes it more likely to reach the limit. We should explore alternatives, to see if we can achieve the same without any additional checkboxes, and without introducing any performance or other regression.

I am not sure whether when this feature was added back in D6 `#states` was a thing or not either, but I don't see any mention of `max_input_vars` in the issues where this was discussed and implemented.

**Comment by yorkshire-pudding** [comment]
I've been doing some experimentation whilst looking at changes needed to Filter Permissions as a result of #980 and I don't think this is an issue for `max_input_vars`; it's hard to say exactly without doing some more detailed debugging, but setting `max_input_vars` slightly over the number of checked boxes seems to clear it.  I'd previously thought it would be the number of permissions so the number of real boxes, but I can drop it much lower and it still works.  I did a search within view source for `checked="checked"` and tried around that number.  It seemed to be the number of checked boxes plus any that have been unchecked as part of the edit, but I can't say for certain.

**Comment by quicksketch** [comment]
> Consider this: on a site with enough roles and permissions to reach 2000 checkboxes on that page, we are adding another 2000 "dummy" checkboxes on top of that, which makes the total number of inputs on the page double

In addition to what @yorkshire-pudding said above; I also don't think the dummy checkboxes have any effect on `max_input_vars` one way or the other. In order to count toward the `max_input_vars` limit an individual checkbox must:

* Have the `checked` property set.
* Have a `name` attribute.
* And not have the `disabled` attribute set.

Dummy checkboxes are both disabled and they don't have a `name` attribute. So when the form is submitted, they add no additional data to POST. You can confirm what gets set in POST by using the network inspector in the browser and submitting the permissions form.

![image](https://user-images.githubusercontent.com/100206/212569382-ef683a34-2d0a-4b37-84bc-c58b0274cb21.png)

So to further clarify when the `max_input_vars` becomes a problem, the limit isn't the total number of checkboxes, it's the total number of _checked_ and _enabled_ checkboxes.

**Comment by quicksketch** [comment]
> adding a hidden form field of `'#type' => 'value',` and storing there the existing permissions, then comparing the form to that in order to determine what has changed.

One solution that could completely side-step the `max_input_vars` limit could use a similar approach: Use JS to accumulate all the checked values into a single field (or maybe one value per role). This could make it so that instead of submitting `# of roles` x `# of permissions`, we would just submit one POST value per role.

I think this would avoid any realistic limits on the number of permissions, but it would only work if JS is enabled. And it would make me nervous manipulating the permissions page in a way that we cannot test (since our automated testing doesn't execute JS).

**Comment by klonos** [comment]
Trying to move this a bit further (or to close it as a non-issue), I've done some more research and, as @kiamlaluno has pointed out already in an earlier comment, checkboxes need a `name` attribute in order to be included in the POST, and the fake checkboxes do not have that attribute set (so they won't affect the `max_input_vars`, which is the main reason for opening this issue here). I guess that that's what this phrase is trying to convey in the inline comment: `We use dummy checkboxes ... because new checkboxes don't alter the submitted form.`. Perhaps we need to elaborate slightly more in that inline comment, and then close this issue here (?).

Before knowing the above, I have searched and found JS-based and CSS-based solutions that render non-`<input>` checkboxes on the page, which we could use in order to replace those HTML dummy checkbox `<input>`s (such as [this](https://stackoverflow.com/questions/40929876/css-checkbox-without-input-tag-onclick-pure-js) for instance), however, I am also nervous about using a JS solution (although it needs to be said that global stats show that the percentage of people having JS disabled in their browsers is very marginal and has been like that for years/decades). I would very much rather we considered something like #6187 instead.

**Comment by klonos** [comment]
I've been doing some more research, and I have found another couple of alternatives that we could consider, both of which would allow us to ditch the duplicate dummy checkboxes:

## tristate checkboxes

Besides the default checked/unchecked states, checkboxes support a third `indeterminate` state. That state applies to radios as well, and it has really good browser support, especially for checkboxes (see https://caniuse.com/?search=indeterminate). So the good news is that we could potentially use this `indeterminate` state to denote that a permission is "inherited" from the Authenticated role:
> A checkbox in the indeterminate state has a horizontal line in the box (it looks somewhat like a hyphen or minus sign) instead of a check/tick in most browsers.

> The indeterminate state is visual only. The checkbox is still either checked or unchecked as a state.

> The control is never a true tri-state control, even if the element's `indeterminate` IDL attribute is set to true. The `indeterminate` IDL attribute only gives the appearance of a third state.

The (perhaps?) not-so-good news:
> You can’t make a checkbox indeterminate through HTML. There is no indeterminate attribute. It is a property of checkboxes though, which you can change via JavaScript.

Why the above is not so good? ...for the same reason @quicksketch mentioned in his previous comment in this thread:
> ...but it would only work if JS is enabled. And it would make me nervous manipulating the permissions page in a way that we cannot test (since our automated testing doesn't execute JS).

Resources:
  - https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/examples/checkbox-mixed
  - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes
  - https://css-tricks.com/indeterminate-checkboxes
  - https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type%3Dcheckbox)

Although not required, there exist some jQuery plugins that enhance tristate checkboxes in various ways:
- https://vanderlee.github.io/tristate
- https://jlbruno.github.io/jQuery-Tristate-Checkbox-plugin

## Custom `value` attribute
As per https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#value , checkboxes support custom values. Here's the interesting bits:

> A string representing the value of the checkbox. This is not displayed on the client-side, but on the server this is the `value` given to the data submitted with the checkbox's `name`. ...
>
> In this example, we've got a name of `subscribe`, and a value of `newsletter`. When the form is submitted, the data name/value pair will be `subscribe=newsletter`.
> 
> If the value attribute was omitted, the default value for the checkbox is `on`, so the submitted data in that case would be `subscribe=on`.

In other words, we could use `value="explicit"` for permission checkboxes that have been manually and explicitly ticked, or `value="inherited"` if the checkbox has been automatically ticked after ticking the checkbox for the respective Anonymous permission. Then in `user_admin_permissions_submit()` we could exclude all checkboxes in the form that had a value of `inherited` (which is basically the same thing we are doing now).

Curious to see what others think about these two options. I'll make sure to mention this in our upcoming dev meeting.
